src/Parser.js

```diff
@@ -807,6 +810,52 @@ export default class Parser {
         return firstToken.range(lastToken, str);
     }
 
+    /**
+     * Parses a group, essentially returning the string formed by the
+     * brace-enclosed tokens plus some position information, possibly
+     * with nested braces.
+     *
+     * @param {string} modeName  Used to describe the mode in error messages
+     * @param {boolean=} optional  Whether the group is optional or required
+     * @return {?Token}
+     */
+    parseStringGroupWithBalancedBraces(modeName, optional) {
+        if (optional && this.nextToken.text !== "[") {
+            return null;
+        }
+        const outerMode = this.mode;
+        this.mode = "text";
+        this.expect(optional ? "[" : "{");
+        let str = "";
+        let nest = 0;
+        const firstToken = this.nextToken;
+        let lastToken = firstToken;
+        while (nest > 0 || this.nextToken.text !== (optional ? "]" : "}")) {
+            if (this.nextToken.text === "EOF") {
+                throw new ParseError(
+                    "Unexpected end of input in " + modeName,
+                    firstToken.range(this.nextToken, str));
+            }
+            lastToken = this.nextToken;
+            str += lastToken.text;
+            if (lastToken.text === "{") {
+                nest += 1;
+            } else if (lastToken.text === "}") {
+                if (nest <= 0) {
+                    throw new ParseError(
+                        "Unbalanced brace of input in " + modeName,
+                        firstToken.range(this.nextToken, str));
+                } else {
+                    nest -= 1;
+                }
+            }
+            this.consume();
+        }
+        this.mode = outerMode;
+        this.expect(optional ? "]" : "}");
+        return firstToken.range(lastToken, str);
```

Comment:
I understand why we're tracking `lastToken` now and why a recursive solution isn't as easy as I thought it might be.